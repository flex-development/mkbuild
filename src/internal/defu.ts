/**
 * @file Internal - defu
 * @module mkbuild/internal/defu
 */

import {
  isUndefined,
  type ExactOptionalPropertyTypes,
  type ObjectPlain
} from '@flex-development/tutils'
import { mergeAndCompare as merge, type Merge } from 'merge-anything'

/**
 * Assigns default properties to the given `object`.
 *
 * Leftmost arguments have more priority when assigning defaults.
 *
 * @template T - Destination object type
 * @template D - Source object type(s)
 *
 * @param {T} object - Destination object
 * @param {D} defaults - Object(s) used to assign default properties
 * @return {ExactOptionalPropertyTypes<Merge<T, D> & T>} `object` with defaults
 */
function defu<T extends ObjectPlain, D extends ObjectPlain[]>(
  object: T,
  ...defaults: D
): ExactOptionalPropertyTypes<Merge<T, D> & T> {
  return merge<T, D>(
    /**
     * Determines how to assign a property.
     *
     * @param {unknown} v1 - Original value from {@linkcode object}
     * @param {unknown} v2 - Incoming value from object in {@linkcode defaults}
     * @return {unknown} Property value for {@linkcode object}
     */
    (v1: unknown, v2: unknown): unknown => (isUndefined(v1) ? v2 : v1),
    object,
    ...defaults
  )
}

export default defu
