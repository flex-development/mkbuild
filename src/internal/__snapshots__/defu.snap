// Vitest Snapshot v1

exports[`unit:internal/defu > should return object with defaults assigned 1`] = `
{
  "bundle": false,
  "clean": true,
  "cwd": ".",
  "dts": true,
  "entries": [
    {
      "ignore": [
        "cli.ts",
      ],
    },
    {
      "bundle": true,
      "minify": true,
      "platform": "node",
      "source": "src/cli.ts",
    },
  ],
  "ext": ".mjs",
  "format": "esm",
  "fs": {
    "lstat": [Function lstat],
    "mkdir": [Function mkdir],
    "readdir": [Function readdir],
    "readdirSync": [Function readdirSync],
    "rm": [Function rm],
    "stat": [Function stat],
    "unlink": [Function unlink],
    "writeFile": [Function writeFile],
  },
  "ignore": Set {
    "**/**.mdx",
    "**/**.stories.*",
    "**/.*ignore",
    "**/.*rc",
    "**/.DS_Store",
    "**/.env*",
    "**/.git*",
    "**/__mocks__/**",
    "**/__snapshots__/**",
    "**/__tests__/**",
    "**/coverage/**",
    "**/package-lock.json",
    "**/yarn.lock",
  },
  "outdir": "dist",
  "pattern": "**",
  "source": "src",
  "sourcemap": true,
  "sourcesContent": false,
  "target": "node14.21.2",
  "tsconfig": "tsconfig.build.json",
}
`;
